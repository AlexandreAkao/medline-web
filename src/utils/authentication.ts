import { isEmpty } from 'ramda';

export const getUserInfo = () => {
  const user = JSON.parse(localStorage.getItem('user') ?? '{}');

  return isEmpty(user) ? null : (user as IUser);
};

export const geToken = () => localStorage.getItem('token');
