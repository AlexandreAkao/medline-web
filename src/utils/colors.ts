export const getColorAlpha = (color: string, alpha = 1) => {
  const alphaLevel = Math.round(255 * alpha).toString(16);
  return color + alphaLevel;
};
