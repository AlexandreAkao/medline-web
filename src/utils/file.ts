export const downloadFile = (blob: Blob, filename: string) => {
  const link = document.createElement('a');
  link.href = URL.createObjectURL(blob);
  link.download = filename;
  link.click();
};
