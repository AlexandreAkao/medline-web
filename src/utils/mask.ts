export enum Mask {
  Empyt = '',
  CPF = '999.999.999-99',
  Date = '99/99/9999',
}
