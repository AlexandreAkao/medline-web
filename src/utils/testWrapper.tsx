import { render } from '@testing-library/react';
import React from 'react';
import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';

import { AuthProvider } from 'providers/AuthProvider';

export const renderWithRouter = (
  component: React.ReactNode,
  { route = '/', history = createMemoryHistory({ initialEntries: [route] }) } = {},
) => {
  return {
    ...render(
      <Router location={history.location} navigator={history}>
        {component}
      </Router>,
    ),
    history,
  };
};

export const renderWithProvider = (
  component: React.ReactNode,
  { route = '/', history = createMemoryHistory({ initialEntries: [route] }) } = {},
) => {
  return {
    ...render(
      <Router location={history.location} navigator={history}>
        <AuthProvider>{component}</AuthProvider>,
      </Router>,
    ),
    history,
  };
};
