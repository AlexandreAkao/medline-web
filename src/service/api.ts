import axios, { AxiosError } from 'axios';
import { toast } from 'react-toastify';

import { getToken } from 'utils/authentication';

const api = axios.create({
  baseURL: (import.meta.env.VITE_BFF_BASE_URL as string) ?? '',
});

api.interceptors.request.use(
  config => {
    const authorization = getToken() ? { Authorization: `Bearer ${getToken()}` } : null;
    return {
      ...config,
      headers: {
        ...authorization,
      },
    };
  },
  error => {
    return Promise.reject(error);
  },
);

api.interceptors.response.use(
  response => {
    return response;
  },
  (error: AxiosError) => {
    if (error.response?.status === 401) {
      axios
        .post<IRefreshToken>(
          'https://securetoken.googleapis.com/v1/token',
          {
            grant_type: 'refresh_token',
            refresh_token: localStorage.getItem('refresh_token') ?? '',
          },
          {
            params: {
              key: 'AIzaSyDYolHTkT0zuI3X1MNup36eSi2gZ3INa2g',
            },
          },
        )
        .then(data => {
          const { access_token: accessToken } = data.data;

          localStorage.setItem('token', accessToken);
        });
    } else if (error.response?.status !== 404)
      if (error.response && error.response.status >= 500)
        toast.error(error.message, { autoClose: 3000, theme: 'colored' });
    return Promise.reject(error);
  },
);

export default api;
