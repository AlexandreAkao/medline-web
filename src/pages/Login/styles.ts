import styled from 'styled-components';

import { TextInputContainer } from 'components/TextInput/styles';
import colors from 'styles/colors';
import doctors from 'assets/images/doctors.svg';
import ScreenSizes from 'styles/screenSizes';

export const LoginContainer = styled.div`
  background-color: ${colors.primary.normal};
  height: 100%;
  overflow: auto;
`;

export const LoginContent = styled.div`
  display: flex;

  > * {
    flex: 1;
    justify-content: center;
  }
`;

export const LoginBackground = styled.div`
  background-image: url(${doctors});
  background-repeat: no-repeat;
  background-position: center;
  height: 600px;

  @media (max-width: ${ScreenSizes.xLarge}) {
    display: none;
  }
`;

export const LoginTitle = styled.h1`
  color: ${colors.white.normal};
  font-size: 1.75rem;
  font-weight: 500;
  margin-bottom: 130px;
`;

export const LoginFormContainer = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  margin: 100px 100px 0 100px;

  @media (max-width: ${ScreenSizes.xLarge}) {
    margin: 50px 50px 0 50px;
  }

  @media (max-width: ${ScreenSizes.medium}) {
    margin: 10px 10px 0 10px;
  }
`;

export const LoginFormLogin = styled.div`
  width: 100%;

  ${TextInputContainer} {
    margin-bottom: 30px;
  }
`;

export const LoginForgotPassword = styled.span`
  color: ${colors.white.normal};
  font-size: 1.125rem;
  font-weight: 500;
`;

export const LoginRemember = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
  margin-bottom: 150px;
`;

export const LoginFormButtons = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
`;
