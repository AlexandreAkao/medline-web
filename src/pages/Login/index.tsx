import { FaUserAlt, FaLock } from 'react-icons/fa';
import { ChangeEvent, useState } from 'react';

import {
  LoginBackground,
  LoginContainer,
  LoginContent,
  LoginForgotPassword,
  LoginFormButtons,
  LoginFormContainer,
  LoginFormLogin,
  LoginRemember,
  LoginTitle,
} from 'pages/Login/styles';
import MedlineLogo from 'assets/images/medline-logo.svg';
import Header from 'components/Header';
import TextInput from 'components/TextInput';
import Checkbox from 'components/Checkbox';
import Button from 'components/Button';
import api from 'service/api';

function Login() {
  const [email, setEmail] = useState<string>('');
  const [password, setPassword] = useState<string>('');

  const handleChangeEmail = (event: ChangeEvent<HTMLInputElement>) => {
    setEmail(event.target.value);
  };

  const handleChangePassword = (event: ChangeEvent<HTMLInputElement>) => {
    setPassword(event.target.value);
  };

  const handleLogin = () => {
    api.post('login', {
      email,
      password,
    });
  };

  return (
    <LoginContainer>
      <Header icon={MedlineLogo} iconAlt="Medline logo" isAuthenticated={false} />

      <LoginContent>
        <LoginBackground />

        <LoginFormContainer>
          <LoginFormLogin>
            <LoginTitle>Entre com seu acesso</LoginTitle>

            <TextInput
              label="UsuÃ¡rio"
              Icon={FaUserAlt}
              placeholder="Email"
              onChange={handleChangeEmail}
              value={email}
            />
            <TextInput
              label="Senha"
              Icon={FaLock}
              isPassword
              placeholder="Senha"
              onChange={handleChangePassword}
              value={password}
            />

            <LoginRemember>
              <Checkbox>Lembrar de mim</Checkbox>
              <LoginForgotPassword>Esqueceu a senha?</LoginForgotPassword>
            </LoginRemember>

            <LoginFormButtons>
              <Button primary={false} size="large" onClick={handleLogin}>
                Login
              </Button>
              <Button primary={false} size="large">
                Cadastrar
              </Button>
            </LoginFormButtons>
          </LoginFormLogin>
        </LoginFormContainer>
      </LoginContent>
    </LoginContainer>
  );
}

export default Login;
