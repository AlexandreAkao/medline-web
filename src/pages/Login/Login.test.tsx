import { screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { vi } from 'vitest';

import { renderWithProvider } from 'utils/testWrapper';
import Login from 'pages/Login';
import api from 'service/api';

describe('Login', () => {
  const renderComponent = () => {
    return renderWithProvider(<Login />);
  };

  it('should match with snapshot', () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  it('should render correctly', () => {
    renderComponent();

    const titleElement = screen.getByText('Entre com seu acesso');
    const checkboxElement = screen.getByRole('checkbox');
    const buttonElements = screen.getAllByRole('button');
    const inputElements = screen.getAllByTestId('test-input');

    expect(titleElement).toBeInTheDocument();
    expect(inputElements).toHaveLength(2);
    expect(checkboxElement).toBeInTheDocument();
    expect(buttonElements).toHaveLength(3);
  });

  it('should send request when click on login button', async () => {
    const post = vi.spyOn(api, 'post').mockImplementation(() =>
      Promise.resolve({
        status: 200,
        data: {
          idToken: 'any_token',
        },
      }),
    );
    vi.spyOn(api, 'get').mockImplementation(() =>
      Promise.resolve({
        status: 200,
        data: {},
      }),
    );
    renderComponent();

    const username = 'any_username';
    const password = 'any_password';

    const inputElements = screen.getAllByTestId('test-input');
    const loginButton = screen.getByText('Login');

    await userEvent.type(inputElements[0], username);
    await userEvent.type(inputElements[1], password);

    await userEvent.click(loginButton);

    expect(post).toBeCalled();
  });

  it('should redirect to register page', async () => {
    const { history } = renderComponent();

    const registerElement = screen.getByText('Cadastrar');
    await userEvent.click(registerElement);

    expect(registerElement).toBeInTheDocument();

    expect(history.location.pathname).toBe('/register');
  });
});
