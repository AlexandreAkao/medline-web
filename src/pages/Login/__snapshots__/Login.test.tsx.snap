// Vitest Snapshot v1

exports[`Login > should match with snapshot 1`] = `
<div>
  <div
    class="sc-eCYdqJ kKlpka"
  >
    <header
      class="sc-fnykZs goXKXN"
      data-testid="test-navigation"
    >
      <img
        alt="Medline logo"
        class="sc-fEOsli bGitPF"
        src="/src/assets/images/medline-logo.svg"
      />
      <nav
        class="sc-bjUoiL dEUTdv"
        data-testid="test-navigation-items"
      >
        <div
          class="sc-hAZoDl cJByLp"
        >
          <button
            class="sc-ksZaOG ioDaHZ"
          >
            Cadastro
          </button>
        </div>
        <div
          class="sc-hAZoDl cJByLp"
        >
          <a
            href="/login"
          >
            Entrar
          </a>
        </div>
      </nav>
    </header>
    <div
      class="sc-jSMfEi clAKAo"
    >
      <div
        class="sc-gKXOVf cGYLZS"
      />
      <div
        class="sc-ftvSup byVrQV"
      >
        <div
          class="sc-papXJ hBdJDP"
        >
          <h1
            class="sc-iBkjds jgQARk"
          >
            Entre com seu acesso
          </h1>
          <div
            class="sc-bczRLJ itlDIl"
          >
            <span
              class="sc-gsnTZi jYHEuj"
              data-testid="test-input-label"
            >
              Usu√°rio
            </span>
            <div
              class="sc-dkzDqf bMJUCZ"
            >
              <svg
                fill="currentColor"
                height="30"
                role="img"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 512 512"
                width="30"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M256 288c79.5 0 144-64.5 144-144S335.5 0 256 0 112 64.5 112 144s64.5 144 144 144zm128 32h-55.1c-22.2 10.2-46.9 16-72.9 16s-50.6-5.8-72.9-16H128C57.3 320 0 377.3 0 448v16c0 26.5 21.5 48 48 48h416c26.5 0 48-21.5 48-48v-16c0-70.7-57.3-128-128-128z"
                />
              </svg>
              <input
                class="sc-hKMtZM fIniom"
                data-testid="test-input"
                placeholder="Email"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class="sc-bczRLJ itlDIl"
          >
            <span
              class="sc-gsnTZi jYHEuj"
              data-testid="test-input-label"
            >
              Senha
            </span>
            <div
              class="sc-dkzDqf bMJUCZ"
            >
              <svg
                fill="currentColor"
                height="30"
                role="img"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 448 512"
                width="30"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M400 224h-24v-72C376 68.2 307.8 0 224 0S72 68.2 72 152v72H48c-26.5 0-48 21.5-48 48v192c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V272c0-26.5-21.5-48-48-48zm-104 0H152v-72c0-39.7 32.3-72 72-72s72 32.3 72 72v72z"
                />
              </svg>
              <input
                class="sc-hKMtZM fIniom"
                data-testid="test-input"
                placeholder="Senha"
                type="password"
                value=""
              />
              <svg
                data-testid="test-eye-icon"
                fill="currentColor"
                height="26"
                role="img"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 576 512"
                width="26"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M572.52 241.4C518.29 135.59 410.93 64 288 64S57.68 135.64 3.48 241.41a32.35 32.35 0 0 0 0 29.19C57.71 376.41 165.07 448 288 448s230.32-71.64 284.52-177.41a32.35 32.35 0 0 0 0-29.19zM288 400a144 144 0 1 1 144-144 143.93 143.93 0 0 1-144 144zm0-240a95.31 95.31 0 0 0-25.31 3.79 47.85 47.85 0 0 1-66.9 66.9A95.78 95.78 0 1 0 288 160z"
                />
              </svg>
            </div>
          </div>
          <div
            class="sc-kDDrLX uIVyf"
          >
            <div
              class="sc-breuTD jsJxhM"
            >
              <input
                class="sc-crXcEl eQIjlf"
                name=""
                type="checkbox"
              />
              <label
                class="sc-evZas fIBvGt"
              >
                Lembrar de mim
              </label>
            </div>
            <span
              class="sc-jqUVSM iwlUUS"
            >
              Esqueceu a senha?
            </span>
          </div>
          <div
            class="sc-iqcoie dLlhNV"
          >
            <button
              class="sc-ksZaOG dHWQgM"
            >
              Login
            </button>
            <button
              class="sc-ksZaOG dHWQgM"
            >
              Cadastrar
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
  ,
</div>
`;
